doc/espnet2_training_option.md

```diff
@@ -0,0 +1,209 @@
+    python -m espnet.bin.asr_train --batch_size 10 --ngpu 2  # Actual batch_size is 20 and each GPU devices are assigned to 10
+    ```
+- ESPnet2: The batch-size is not changed regardless of the number of GPUs.
+    - Therefore, you should more number of batch-size than the number of GPUs.
+    ```bash
+    python -m espnet.bin.asr_train --batch_size 10 --ngpu 2  # Actual batch_size is 10 and each GPU devices are assigned to 5
+    ```
+
+Note that even espnet1, if using `bin` or `frame` batch-count, this changing of batch_size is not done.
+
+## Gradient accumulating
+There are several measures to deal large model architecture than your GPU device memory when training.
+
+- Using more number of GPUs
+- Using half decision tensor
+- Using [torch.utils.checkpoint](https://pytorch.org/docs/stable/checkpoint.html)
+- Gradient accumulating
+
+Gradient accumulating is a technique to handle larger mini-batch than available size.
+
+Split a mini-batch into several numbers and forward and backward for each pieces and accumulate the gradients, while optimizer's updating is invoked every the number of forwarding just like following:
+
+```python
+# accum_grad is the number of pieces
+for i, batch in enumerate(iterator):
+    loss = net(batch)
+    (loss / accum_grad).backward()  # Gradients are accumulated
+    if i % accum_grad:
+        optim.update()
+        optim.zero_grads()
+```
+
+Give `--accum_grad <int>` to use this option.
+
+```bash
+python -m espnet.bin.asr_train --accum_grad 2
+```
+
+ The effective batch_size becomes **almost** same as `accum_grad * batch_size` except for:
+
+- The random state
+- Some statistical layers based on mini-batch e.g. BatchNormalization
+- The case that the batch_size is not unified for each iterations.
+
+## Using apex: mixed precision training
+See also: https://github.com/NVIDIA/apex
+
+```bash
+python -m espnet.bin.asr_train --train_dtype float16  # Just training with half precision
+python -m espnet.bin.asr_train --train_dtype float32  # default
+python -m espnet.bin.asr_train --train_dtype float64
+python -m espnet.bin.asr_train --train_dtype O0  # opt_level of apex
+python -m espnet.bin.asr_train --train_dtype O1  # opt_level of apex
+python -m espnet.bin.asr_train --train_dtype O2  # opt_level of apex
+python -m espnet.bin.asr_train --train_dtype O3  # opt_level of apex
+```
+
+## Reprodusibility and determinism
+There are some possibilities to make training not-reproducible.
+
+- Initialization of parameters which comes from pytorch/espnet version difference.
+- Reducing order for float values when multi GPUs
+    - I don't know whether NCCL is deterministic or not.
+- Random seed difference
+    - We fixed the random seed for each epochs, so the randomness should be reproduced even if the process is resumed.
+- CuDNN or some non-deterministic operations for CUDA: See https://pytorch.org/docs/stable/notes/randomness.html
+
+By default, CuDNN performs non-deterministic mode and it can be changed by:
```

Comment:
I thought that we perform `--cudnn_deterministic true` as default.
Could you double-check it?
It's better to use deterministic as default.

Reply:
Deterministic mode shouldn't be true by default because it might be lower performance than non-deterministic.

https://pytorch.org/docs/stable/notes/randomness.html#cudnn

Reply:
Yes, but I want to stick to make it deterministic as much as possible for the following reasons.

- Many researchers want to have deterministic implementations for fair comparisons. I know that it is not deterministic due to the GPU, GPU driver, CUDA, CUDNN etc., but it still reproduces the same result with a specific condition
- It is very hard to debug it without the deterministic mode (the deterministic mode with the fixed seed assures that the forward computation is the same at least).
- If people wrongly set the non-deterministic mode in the inference, it would be more serious (ASR/TTS results would be changed every trial).

Could you reconsider to make the deterministic mode default?
(I also once tested the non-deterministic and deterministic mode and I could not observe so much difference).

Reply:
O.K. I'll follow you though I still hesitate because the performance difference is dependent on the model architecture and GPU device. ESpnet happens not to use deep cnn arch, so this was not big issue.

In my small benchmark,  it's 1% slower for resnet152 with Geforce RTX 2080ti, v100, tesla-k80, Geforce 1080ti. This is acceptable for me.

```python
import torch
from torchvision.models import resnet

def main():
    print(torch.backends.cudnn.version())
    torch.manual_seed(0)
    conv = resnet.resnet152().cuda()
    x = torch.randn(30, 3, 224, 224).cuda()
    ntime = 20

    for benchmark in [True, False]:
        for deterministic in [False, True]:
            with torch.backends.cudnn.flags(
                deterministic=deterministic, benchmark=benchmark, verbose=True,
            ):
                start_event = torch.cuda.Event(enable_timing=True)
                end_event = torch.cuda.Event(enable_timing=True)

                start_event.record()
                for i in range(ntime):
                    y = conv(x)
                    loss = y.mean()
                    loss.backward()
                    conv.zero_grad()
                    if i == 0:
                        end_event.record()
                        torch.cuda.synchronize()
                        first = start_event.elapsed_time(end_event)
                        start_event.record()

                    end_event.record()
                    torch.cuda.synchronize()
                    times = start_event.elapsed_time(end_event)

                print(
                    f"benchmark={benchmark}, deterministic={deterministic}: "
                    f"{times / ntime:.4e}[s/iter] "
                    f"(at the first iteration: {first:.4e}[s/iter])"
                )
main()
```
```
7603
benchmark=True, deterministic=False: 5.9307e+02[s/iter] (at the first iteration: 9.6397e+02[s/iter])
benchmark=True, deterministic=True: 6.0100e+02[s/iter] (at the first iteration: 6.2919e+02[s/iter])
benchmark=False, deterministic=False: 6.0787e+02[s/iter] (at the first iteration: 6.3496e+02[s/iter])
benchmark=False, deterministic=True: 6.1303e+02[s/iter] (at the first iteration: 6.4359e+02[s/iter])
```

Reply:
I changed determinic mode to default